File: src/main/java/guru/nidi/brainfuck/Opcodes.java
/*
 * Copyright © 2017 Stefan Niederhauser (nidin@gmx.ch)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.nidi.brainfuck;

class Opcodes {
    //jumps could be optimized with short jump if possible
    static final byte[]
            DEC_PTR = new byte[]{0x4e}, //dec esi
            INC_PTR = new byte[]{0x46}, //inc esi
            DEC_VAL = new byte[]{(byte) 0xfe, 0x0e}, //dec byte[esi]
            INC_VAL = new byte[]{(byte) 0xfe, 0x06}, //inc byte[esi]
            JZ = new byte[]{(byte) 0x80, 0x3e, 0x00, 0x0f, (byte) 0x84}, //cmp byte[esi],0; je + 4 bytes relative
            JUMP = new byte[]{(byte) 0xe9}, //jmp + 4 bytes relative
            CALL = new byte[]{(byte) 0xe8}, //call + 4 bytes relative dest
            RET = new byte[]{(byte) 0xc3}; //ret
}

File: src/main/kotlin/guru/nidi/brainfuck/Opcodes.kt
/*
 * Copyright © 2017 Stefan Niederhauser (nidin@gmx.ch)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package guru.nidi.brainfuck

internal fun bytes(vararg bs: Int) = ByteArray(bs.size).also {
    bs.forEachIndexed { value, index -> bs[index] = value }
}

val incEcx = bytes(0x41)
val decEcx = bytes(0x49)

fun addEcxConst1(value: Int) = bytes(0x83, 0xc1, value)
fun subEcxConst1(value: Int) = bytes(0x83, 0xe9, value)
fun addEcxConst4(value: Int) = bytes(0x81, 0xc1, value, value shr 8, value shr 16, value shr 24)
fun subEcxConst4(value: Int) = bytes(0x81, 0xe9, value, value shr 8, value shr 16, value shr 24)

val inc1PtrEcx = bytes(0xfe, 0x01)
val dec1PtrEcx = bytes(0xfe, 0x09)
fun add1PtrEcx(value: Int) = bytes(0x80, 0x01, value)
fun sub1PtrEcx(value: Int) = bytes(0x80, 0x29, value)

val callPtrEdi = bytes(0xff, 0x17)
val callPtrEsi = bytes(0xff, 0x16)

val ret = bytes(0xc3)

fun cmp1PtrEcx(value: Int) = bytes(0x80, 0x3e, value)

fun jump1IfZero(value: Int) = bytes(0x74, value)
//TODO is there jump2?
fun jump4IfZero(value: Int) = bytes(0x0f, 0x84, value, value shr 8, value shr 16, value shr 24)

fun jump1(value: Int) = bytes(0xeb, value)
//TODO is there jump2?
fun jump4(value: Int) = bytes(0xe9, value, value shr 8, value shr 16, value shr 24)
